src/jinja2/filters.py

```diff
@@ -218,6 +218,34 @@ def do_lower(s: str) -> str:
     return soft_str(s).lower()
 
 
+def do_pairs(value: t.Union[t.Mapping[K, V], Undefined]) -> t.Iterable[t.Tuple[K, V]]:
+    """Returns an iterator over the pairs of a mapping in the order that
+    the mapping's iterator returns.  In practical terms ``x|pairs`` is
+    the same as ``x.items()`` with two differences.  The first is that
+    if ``x`` is undefined an empty iterator is returned, secondly the
+    filter is a more portable version to alternative Jinja implementations
+    which might not have a ``.items()`` method on maps.
+
+    Example usage:
+
+    .. sourcecode:: html+jinja
+
+        <dl>
+        {% for key, value in my_dict|pairs %}
+            <dt>{{ key }}
+            <dd>{{ value }}
+        {% endfor %}
+        </dl>
+
+    .. versionadded:: 3.1
+    """
+    if isinstance(value, Undefined):
+        return
+    if not isinstance(value, abc.Mapping):
+        raise TypeError("Can only get pairs from mappings")
+    yield from iter(value.items())
```

Comment:
Can't we avoid what you're talking about with something like this:
```suggestion
    if isinstance(value, abc.Mapping):
        return value.items()
    elif isinstance(value, abc.ItemsView):
        return value
    else:
        raise TypeError(f"Expected Mapping or ItemsView, got {type(value)}")
```

Also I think you have a superfluous `iter` call in your current implementation: instead of `yield from iter(value.items())` you can just `yield from value.items()` (though better yet just `return value.items()` directly as in the above)?

Reply:
It seems like it would be fairly uncommon for a user to pass an `ItemsViiew`, and that's probably hiding some other issue with their data or template. I'd rather keep this simple and only support `Mapping`.

I do agree that `return value.items()`, as well as your error message, seem preferable.

Reply:
Makes sense. In that case I vote for calling this `items` rather than `pairs`.

Reply:
I changed it to `items`. The `return value.items()` doesn't work because of the undefined branch unless that gets changed to `return iter(())` which seems odd.

Reply:
Thanks for the changes.

I actually think that making the undefined branch `return iter(())` would make the code more clearly and correctly match the docstring you've written (you no longer have to keep reading the code to see if a `yield` statement appears somewhere further down to make sure this branch works correctly).

Relatedly, might be worth adding a test covering the undefined branch.

Reply:
I don't agree on the code path changes. I am also not a fan of leaking through types if I don't have to. `.items()` doesn't actually return an iterator but a view and that's not part of the API contract. I did add a test for undefined iteration however.

Reply:
Thanks for adding the tests. Fine with me to merge. @davidism?

Regarding

> .items() doesn't actually return an iterator but a view and that's not part of the API contract

The signature you've written is
```python
def do_items(value: t.Union[t.Mapping[K, V], Undefined]) -> t.Iterable[t.Tuple[K, V]]:
```
i.e. you're promising an Iterable, not an Iterator, and
```python
>>> issubclass(collections.abc.ItemsView, collections.abc.Iterable)
True
```
so returning an ItemsView here would be fulfilling this API contract.
```python
# demo.py
import typing as t

K = t.TypeVar("K")
V = t.TypeVar("V")

def do_items(value: t.Mapping[K, V]) -> t.Iterable[t.Tuple[K, V]]:
    return {}.items()
```
```console
> mypy demo.py
Success: no issues found in 1 source file
```
Liskov substitution principle blah blah blah.

I agree there are certainly benefits of hiding subtypes from callers. But then I would still prefer `return iter(value.items())` so we avoid the "spooky action at a distance  effect" of the yield statement below the earlier `return` statement changing the meaning.